src/System.Net.Sockets/src/System/Net/Sockets/IOControlKeepAlive.cs

```diff
@@ -0,0 +1,103 @@
+        }
+
+        public static SocketError? Set(SafeSocketHandle handle, SocketOptionLevel optionLevel, SocketOptionName optionName, int optionValue)
+        {
+            if (IsNotNeeded(optionLevel, optionName))
+                return null;
+
+            IOControlKeepAlive ioControlKeepAlive = s_socketKeepAliveCwt.GetOrCreateValue(handle);
+            if (optionName == SocketOptionName.TcpKeepAliveTime)
+                ioControlKeepAlive._timeMs = SecondsToMilliseconds(optionValue);
+            else
+                ioControlKeepAlive._intervalMs = SecondsToMilliseconds(optionValue);
+
+            int realOptionLength = 0;
+            return SocketPal.WindowsIoctl(handle, unchecked((int)IOControlCode.KeepAliveValues), ioControlKeepAlive.ToByteArray(), null, out realOptionLength);
+        }
+
+        private static bool SupportsKeepAliveViaSocketOption()
+        {
+            using (Socket socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                int time = MillisecondsToSeconds(WindowsDefaultTimeMs);
+                SocketError timeErrCode = Interop.Winsock.setsockopt(
+                    socket.SafeHandle,
+                    SocketOptionLevel.Tcp,
+                    SocketOptionName.TcpKeepAliveTime,
+                    ref time,
+                    sizeof(int));
+                int interval = MillisecondsToSeconds(WindowsDefaultIntervalMs);
+                SocketError intervalErrCode = Interop.Winsock.setsockopt(
+                    socket.SafeHandle,
+                    SocketOptionLevel.Tcp,
+                    SocketOptionName.TcpKeepAliveInterval,
+                    ref time,
+                    sizeof(int));
+                return timeErrCode == SocketError.Success && intervalErrCode == SocketError.Success;
+            }
+        }
+
+        private static bool IsNotNeeded(SocketOptionLevel optionLevel, SocketOptionName optionName)
+        {
+            return s_supportsKeepAliveViaSocketOption ||
+                optionLevel != SocketOptionLevel.Tcp ||
+                (optionName != SocketOptionName.TcpKeepAliveTime && optionName != SocketOptionName.TcpKeepAliveInterval);
+        }
+
+        private static int MillisecondsToSeconds(uint milliseconds)
+        {
+            return (int)(milliseconds / 1000u);
+        }
+
+        private static uint SecondsToMilliseconds(int seconds)
+        {
+            return (uint)seconds * 1000u;
+        }
+
+        private byte[] ToByteArray()
+        {
+            const int uintSize = sizeof(uint);
+            const uint onOff = 1u;
+            byte[] buffer = s_keepAliveValuesBuffer ?? (s_keepAliveValuesBuffer = new byte[3 * uintSize]);
+            BitConverter.TryWriteBytes(buffer.AsSpan(), onOff);
+            BitConverter.TryWriteBytes(buffer.AsSpan(uintSize), _timeMs);
+            BitConverter.TryWriteBytes(buffer.AsSpan(uintSize * 2), _intervalMs);
+            return buffer;
```

Comment:
I would prefer the Fill approach rather than this ToByteArray method. It's not at all cost to the caller that this returns a ThreadStatic, which makes it error prone. That's why I'd like the call site to be the one that accesses the array, and then passes it to the helper to be filled.